File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-202259-53404/analysis/acceptance-project/src/main/java/customerimporter/NewZipCode.java
package customerimporter;

import static customerimporter.Validation.assertThat;
import static customerimporter.Validation.isAllDigits;
import static customerimporter.Validation.isAllLetters;

import java.util.function.Supplier;

public class NewZipCode extends ZipCode {

    // "B1636BBE": one letter, the old zip code and three letters
    private static final int SIZE = 8;
    private static final int MINIMUM_OLD_ZIP_CODE_VALUE = 1000;
    private static final int NUMBER_OF_ENDING_LETTERS = 3;

    private final String value;

    // instance creation

    public static boolean canBeCreatedFrom(String aZipCode) {
        return Character.isLetter(aZipCode.charAt(0));
    }

    public static NewZipCode with(String aZipCode) {
        assertIsValid(aZipCode);
        return new NewZipCode(aZipCode);
    }

    // error messages

    public static String invalidZipCodeErrorDescription() {
        return "Invalid new zipcode";
    }

    // asserting - private

    private static void assertIsValid(String aZipCode) {
        assertThat(aZipCode.length() == SIZE, invalidZipCodeErrorDescription());
        assertThat(isAllDigits(oldZipCodePartOf(aZipCode)), invalidZipCodeErrorDescription());
        assertThat(Integer.parseInt(oldZipCodePartOf(aZipCode)) >= MINIMUM_OLD_ZIP_CODE_VALUE, invalidZipCodeErrorDescription());
        assertThat(isAllLetters(endingLettersOf(aZipCode)), invalidZipCodeErrorDescription());
    }

    private static String endingLettersOf(String aZipCode) {
        return aZipCode.substring(aZipCode.length() - NUMBER_OF_ENDING_LETTERS);
    }

    private static String oldZipCodePartOf(String aZipCode) {
        return aZipCode.substring(1, 5);
    }

    // initialization

    private NewZipCode(String aValue) {
        value = aValue;
    }

    // accessing

    public Object value() {
        return value;
    }

    // testing

    public boolean isNew() {
        return true;
    }

    public boolean isOld() {
        return false;
    }

    // evaluating

    public Object newZipCodeIfNone(Supplier<Object> aNoneBlock) {
        return value();
    }

    public Object oldZipCodeIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-203929-73794/analysis/own-tests-project/src/customerimporter/new_zip_code.py
from customerimporter.zip_code import ZipCode


class NewZipCode(ZipCode):

    # type

    @classmethod
    def recognizes(cls, an_imported_zip_code):
        return an_imported_zip_code[:1].isalpha()

    def is_new(self):
        return True

    # zip code

    def new_zip_code_if_none(self, a_none_block):
        return self.value()

    # error messages

    @classmethod
    def invalid_value_error_description(cls):
        return "Invalid new zipcode"

    # validation

    @classmethod
    def size(cls):
        # "B1636BBE" size 8
        return 8

    @classmethod
    def assert_is_valid(cls, an_imported_zip_code):
        cls._assert_that(len(an_imported_zip_code) == cls.size())
        cls._assert_that(cls._is_old_zip_code_number(cls._old_zip_code_part_of(an_imported_zip_code)))
        cls._assert_that(cls._are_all_letters(an_imported_zip_code[-3:]))

    # validation - private

    @classmethod
    def _old_zip_code_part_of(cls, an_imported_zip_code):
        return an_imported_zip_code[1:5]
